### DISCUSSION TITLE
Error when using custom model

### DISCUSSION THREAD

**Discussion by wolfsblu** [janko/rodauth-rails#166] [Q&A] [answered]
Hello! I'm having problems trying to use a custom model with `rodauth-rails`. My setup is as follows: 

I have a custom `ActiveRecord` model called `User`

```ruby
class User < ApplicationRecord
    include Rodauth::Rails.model

    enum :status, unverified: 1, verified: 2, closed: 3
end
```
Then I configured `rodauth-rails` to use the model above in `app/misc/rodauth_main.rb` and adapted the migrations accordingly:
```ruby
class RodauthMain < Rodauth::Rails::Auth
  configure do
    ...
    accounts_table :users 
    rails_account_model User

    login_column :name
    login_uses_email? false

    email_to do 
      account[:email]
    end
end
```
My only existing `RodauthApp` (`app/misc/rodauth_app,rb`) looks as follows:
```ruby
require "rodauth/model"
class RodauthApp < Rodauth::Rails::App
  configure RodauthMain

  route do |r|
    rodauth.load_memory 
    r.on "auth" do
      r.rodauth
    end
  end
end
```
The problem is whenever I try to call a protected endpoint in one of my API controllers I get this error:
```
NoMethodError (undefined method `rodauth' for RodauthApp:Class

        rodauth(name) or fail ArgumentError, "unknown rodauth configuration: #{name.inspect}"
        ^^^^^^^
Did you mean?  rodauth!):

app/models/user.rb:3:in `<class:User>'
app/models/user.rb:2:in `<main>'
app/misc/rodauth_main.rb:13:in `block in <class:RodauthMain>'
app/misc/rodauth_main.rb:2:in `<class:RodauthMain>'
app/misc/rodauth_main.rb:1:in `<main>'
app/misc/rodauth_app.rb:5:in `<class:RodauthApp>'
app/misc/rodauth_app.rb:4:in `<main>'
```

Does anybody have an idea what could cause this error?

**Comment by janko** [comment, marked as answer]
Hi, thanks for posting the relevant code. I believe there is a circular dependency issue going on here:

1. when the Rodauth middleware references `RodauthApp`, it starts loading `rodauth_app.rb`
2. when `rodauth_app.rb` references `RodauthMain`, it starts loading `rodauth_main.rb` (at that point the Rodauth configuration hasn't yet been registered into the app)
3. when `rodauth_main.rb` references the `User` model, it starts loading `user.rb`
4. when `user.rb` includes the model mixin, the mixin tries to retrieve the Rodauth configuration in order to read table & column names
5. since no configuration hasn't yet been registered, the code fails

Ideally, you should have received an error that the main configuration (`nil`) doesn't exist, but you received a `NoMethodError`, because the Rodauth plugin gets loaded into the Roda app only on first `configure` (and it didn't manage).

Changing `rails_account_model` to load the `User` class at runtime should fix the issue (I will update the docs):

```rb
rails_account_model { User }
```

Also, note that setting `rails_account_model` is needed only when the model name cannot be inferred from the table name (typically only when using namespaces). Since the table here is `:users`, it will automatically assume the `User` model class.

**Reply by wolfsblu** [reply]
Interesting, that makes sense. Thanks for the detailed and quick response!
